devtools/src/main/scala/kalix/devtools/impl/KalixProxyContainer.scala

```diff
@@ -0,0 +1,124 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kalix.devtools.impl
+
+import com.typesafe.config.Config
+import kalix.devtools.BuildInfo
+import kalix.devtools.impl.KalixProxyContainer.KalixProxyContainerConfig
+import org.slf4j.LoggerFactory
+import org.testcontainers.Testcontainers
+import org.testcontainers.containers.GenericContainer
+import org.testcontainers.containers.output.Slf4jLogConsumer
+import org.testcontainers.utility.DockerImageName
+
+object KalixProxyContainer {
+
+  object KalixProxyContainerConfig {
+    def apply(config: Config): KalixProxyContainerConfig = {
+
+      // FIXME: core Kalix should be able to read dev-mode.conf from root folder as well
+      // as such, we can kalix.user-function-port configured to something else without touching the prod config
+      val userFunctionPort = config.getInt("kalix.user-function-port")
+      val containerConfig = config.getConfig("kalix.dev-mode.proxy-container")
+
+      KalixProxyContainerConfig(
+        proxyImage = containerConfig.getString("proxy-image"),
+        proxyPort = containerConfig.getInt("proxy-port"),
+        userFunctionPort = userFunctionPort,
+        serviceName = containerConfig.getString("service-name"),
+        aclEnabled = containerConfig.getBoolean("acl-enabled"),
+        viewFeaturesAll = containerConfig.getBoolean("view-features-all"),
+        brokerConfigFile = containerConfig.getString("broker-config-file"),
+        pubsubEmulatorHost = containerConfig.getString("pubsub-emulator-host"))
+
+    }
+  }
+
+  case class KalixProxyContainerConfig(
+      proxyImage: String,
+      proxyPort: Int,
+      userFunctionPort: Int,
+      serviceName: String,
+      aclEnabled: Boolean,
+      viewFeaturesAll: Boolean,
+      brokerConfigFile: String,
+      pubsubEmulatorHost: String)
+
+  val logger = LoggerFactory.getLogger(classOf[KalixProxyContainer])
+
+  def apply(config: KalixProxyContainerConfig): KalixProxyContainer = {
+
+    val dockerImage: DockerImageName =
+      if (config.proxyImage.trim.nonEmpty) {
+        logger.info("Using custom proxy image [{}]", config.proxyImage)
+        DockerImageName.parse(config.proxyImage)
+      } else {
+        DockerImageName.parse(BuildInfo.proxyImage).withTag(BuildInfo.proxyVersion)
+      }
+
+    new KalixProxyContainer(dockerImage, config)
+  }
+
+}
+
+class KalixProxyContainer private (image: DockerImageName, config: KalixProxyContainerConfig)
+    extends GenericContainer[KalixProxyContainer](image) {
+
+  private val containerLogger = LoggerFactory.getLogger("kalix-proxy-server")
+  containerLogger.info("KalixProxyContainer config : {}", config)
+  withLogConsumer(new Slf4jLogConsumer(containerLogger))
+
+  val proxyPort = config.proxyPort
+  val userFunctionPort = config.userFunctionPort
+  addFixedExposedPort(proxyPort, proxyPort)
```

Comment:
testcontainers folks really unrecommend fixing the port. I understand them. They are very focus on testing and in that context, fixed ports is are recipe to conflicts. 

But we need fix ports so users know there to reach the proxy, ie: localhost:9000.

Maybe, we should not use testcontainers here, but https://github.com/docker-java/docker-java directly.